File: platforms/ios/Dezel/Dezel/Modules/Global/JavaScriptXMLHttpRequest.swift
import Foundation

/**
 * @class JavaScriptXMLHttpRequest
 * @since 0.7.0
 */
open class JavaScriptXMLHttpRequest: JavaScriptClass, HttpRequestDelegate {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * @property request
	 * @since 0.7.0
	 * @hidden
	 */
	private var request: HttpRequest?

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method dispose
	 * @since 0.6.0
	 */
	override open func dispose() {

		self.request?.abort()
		self.request = nil

		super.dispose()
	}

	//--------------------------------------------------------------------------
	// MARK: Methods - Request Delegate
	//--------------------------------------------------------------------------

	/**
	 * @method didError
	 * @since 0.7.0
	 * @hidden
	 */
	public func didError(request: HttpRequest, error: NSError) {

		self.holder.callMethod("nativeOnError", arguments: [
			self.context.createNumber(error.code),
			self.context.createString(error.localizedDescription)
		])

		self.unprotect()
	}

	/**
	 * @method didTimeout
	 * @since 0.7.0
	 * @hidden
	 */
	public func didTimeout(request: HttpRequest, error: NSError) {

		self.holder.callMethod("nativeOnTimeout", arguments: [
			self.context.createNumber(error.code),
			self.context.createString(error.localizedDescription)
		])

		self.unprotect()
	}

	/**
	 * @method didProgress
	 * @since 0.7.0
	 * @hidden
	 */
	public func didProgress(request: HttpRequest, loaded: Int64, length: Int64) {
		self.holder.callMethod("nativeOnProgress", arguments: [
			self.context.createNumber(Double(loaded)),
			self.context.createNumber(Double(length))
		])
	}

	/**
	 * @method didComplete
	 * @since 0.7.0
	 * @hidden
	 */
	public func didComplete(request: HttpRequest, response: HttpResponse) {

		self.holder.callMethod("nativeOnComplete", arguments: [
			self.context.createString(response.data),
			self.context.createNumber(response.statusCode),
			self.context.createString(response.statusText),
			self.context.createObject(response.headers),
			self.context.createString(response.url)
		])

		self.unprotect()
	}

	//--------------------------------------------------------------------------
	// MARK: JS Function
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_request
	 * @since 0.7.0
	 * @hidden
	 */
	@objc open func jsFunction_request(callback: JavaScriptFunctionCallback) {

		self.request?.abort()
		self.request = nil

		let requestUrl      = callback.argument(0)
		let requestMethod   = callback.argument(1).string
		let requestHeaders  = callback.argument(2)
		let requestTimeout  = callback.argument(3)
		let requestUsername = callback.argument(4)
		let requestPassword = callback.argument(5)
		let requestData     = callback.argument(6)

		guard let url = requestUrl.toURL() else {
			NSLog("Invalid XMLHttpRequest URL.")
			return
		}

		let headers = requestHeaders.toDictionaryOfString()
		let timeout = requestTimeout.toTimeInterval() / 1000

		var username: String?
		var password: String?

		if (requestUsername.isString) { username = requestUsername.string }
		if (requestPassword.isString) { password = requestPassword.string }

		let request = HttpRequest(url: url, method: requestMethod)
		request.timeout = timeout
		request.headers = headers
		request.username = username
		request.password = password
		request.delegate = self

		if (requestData.isString) {
			request.data = requestData.string.data(using: .utf8)
		}

		request.send()

		self.request = request

		self.protect()
	}

	/**
	 * @method jsFunction_abort
	 * @since 0.7.0
	 * @hidden
	 */
	@objc open func jsFunction_abort(callback: JavaScriptFunctionCallback) {
		self.request?.abort()
	}

	/**
	 * @method jsFunction_reset
	 * @since 0.7.0
	 * @hidden
	 */
	@objc open func jsFunction_reset(callback: JavaScriptFunctionCallback) {
		self.request?.delegate = nil
		self.request?.abort()
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/global/JavaScriptXMLHttpRequest.kt
package ca.logaritm.dezel.modules.global

import android.util.Log
import ca.logaritm.dezel.core.JavaScriptClass
import ca.logaritm.dezel.core.JavaScriptContext
import ca.logaritm.dezel.core.JavaScriptFunctionCallback
import ca.logaritm.dezel.extension.core.createObject
import ca.logaritm.dezel.extension.core.createString
import ca.logaritm.dezel.extension.core.toDictionaryOfString
import ca.logaritm.dezel.extension.core.toURL
import ca.logaritm.dezel.extension.fatalError
import ca.logaritm.dezel.networking.HttpRequest
import ca.logaritm.dezel.networking.HttpRequestListener
import ca.logaritm.dezel.networking.HttpResponse

/**
 * @class JavaScriptXMLHttpRequest
 * @super JavaScriptClass
 * @since 0.7.0
 */
open class JavaScriptXMLHttpRequest(context: JavaScriptContext) : JavaScriptClass(context), HttpRequestListener {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * @property request
	 * @since 0.7.0
	 * @hidden
	 */
	private var request: HttpRequest? = null

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @method dispose
	 * @since 0.6.0
	 */
	override fun dispose() {
		this.request?.abort()
		this.request = null
		super.dispose()
	}

	//--------------------------------------------------------------------------
	// Request JavaScriptViewListener
	//--------------------------------------------------------------------------

	/**
	 * @method onError
	 * @since 0.7.0
	 * @hidden
	 */
	override fun onError(request: HttpRequest, response: HttpResponse) {

		this.callMethod("nativeOnError", arrayOf(
			this.context.createNumber(response.statusCode),
			this.context.createString(response.statusText)
		))

		this.unprotect()
	}

	/**
	 * @method onTimeout
	 * @since 0.7.0
	 * @hidden
	 */
	override fun onTimeout(request: HttpRequest, response: HttpResponse) {

		this.callMethod("nativeOnTimeout", arrayOf(
			this.context.createNumber(response.statusCode),
			this.context.createString(response.statusText)
		))

		this.unprotect()
	}

	/**
	 * @method onProgress
	 * @since 0.7.0
	 * @hidden
	 */
	override fun onProgress(request: HttpRequest, loaded: Int, length: Int) {
		this.callMethod("nativeOnProgress", arrayOf(
			this.context.createNumber(loaded),
			this.context.createNumber(length)
		))
	}

	/**
	 * @method onComplete
	 * @since 0.7.0
	 * @hidden
	 */
	override fun onComplete(request: HttpRequest, response: HttpResponse) {

		this.callMethod("nativeOnComplete", arrayOf(
			this.context.createString(response.data),
			this.context.createString(response.statusText),
			this.context.createNumber(response.statusCode),
			this.context.createObject(response.headers),
			this.context.createString(response.url)
		))

		this.unprotect()
	}

	//--------------------------------------------------------------------------
	// JS funtion
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_request
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_request(callback: JavaScriptFunctionCallback) {

		this.request?.abort()
		this.request = null

		if (callback.arguments < 7) {
			fatalError("Method JavaScriptXMLHttpRequest.request() requires 7 arguments.")
		}

		val requestUrl      = callback.argument(0)
		val requestMethod   = callback.argument(1).string
		val requestHeaders  = callback.argument(2)
		val requestTimeout  = callback.argument(3)
		val requestUsername = callback.argument(4)
		val requestPassword = callback.argument(5)
		val requestData     = callback.argument(6)

		val url = requestUrl.toURL()
		if (url == null) {
			Log.e("DEZEL", "Invalid XMLHttpRequest URL.")
			return
		}

		val headers = requestHeaders.toDictionaryOfString()
		val timeout = requestTimeout.number

		var username: String? = null
		var password: String? = null

		if (requestUsername.isString) { username = requestUsername.string }
		if (requestPassword.isString) { password = requestPassword.string }

		val request = HttpRequest(url, requestMethod)
		request.timeout = timeout
		request.headers = headers
		request.username = username
		request.password = password
		request.listener = this

		if (requestData.isString) {
			request.data = requestData.string.toByteArray(charset("UTF-8"))
		}

		request.send()

		this.request = request

		this.protect()
	}

	/**
	 * @method jsFunction_abort
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_abort(callback: JavaScriptFunctionCallback) {
		this.request?.abort()
	}

	/**
	 * @method jsFunction_reset
	 * @since 0.7.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_reset(callback: JavaScriptFunctionCallback) {
		this.request?.listener = null
		this.request?.abort()
	}	
}